transport/awssqs/consumer.go

```diff
@@ -0,0 +1,296 @@
+// the messages are retrieved from the consumer's configured queue.
+func (c Consumer) Consume(ctx context.Context, receiveMsgInput *sqs.ReceiveMessageInput) error {
+	receiveMsgInput.QueueUrl = &c.queueURL
+	out, err := c.sqsClient.ReceiveMessageWithContext(ctx, receiveMsgInput)
+	if err != nil {
+		return err
+	}
+	return c.handleMessages(ctx, out.Messages)
+}
+
+// handleMessages handles the consumed messages.
+func (c Consumer) handleMessages(ctx context.Context, msgs []*sqs.Message) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Copy received messages slice in leftMsgs slice
+	// leftMsgs will be used by the consumer's visibilityTimeoutFunc to extend the
+	// visibility timeout for the messages that have not been processed yet.
+	leftMsgs := []*sqs.Message{}
+	leftMsgs = append(leftMsgs, msgs...)
+
+	visibilityTimeoutCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go c.visibilityTimeoutFunc(visibilityTimeoutCtx, c.sqsClient, c.queueURL, c.visibilityTimeout, &leftMsgs, c.leftMsgsMux)
+
+	if len(c.finalizer) > 0 {
+		defer func() {
+			for _, f := range c.finalizer {
+				f(ctx, &msgs)
+			}
+		}()
+	}
+
+	for _, f := range c.before {
+		ctx = f(ctx, &msgs)
+	}
+
+	for _, msg := range msgs {
+		if c.deleteMessage == BeforeHandle {
+			if _, err := c.sqsClient.DeleteMessageWithContext(ctx, &sqs.DeleteMessageInput{
+				QueueUrl:      &c.queueURL,
+				ReceiptHandle: msg.ReceiptHandle,
+			}); err != nil {
+				c.errorHandler.Handle(ctx, err)
+				c.errorEncoder(ctx, err, msg, c.sqsClient)
+				return err
+			}
+		}
+
+		if err := c.handleSingleMessage(ctx, msg, &leftMsgs); err != nil {
+			return err
+		}
+
+		if c.deleteMessage == AfterHandle {
+			if _, err := c.sqsClient.DeleteMessageWithContext(ctx, &sqs.DeleteMessageInput{
+				QueueUrl:      &c.queueURL,
+				ReceiptHandle: msg.ReceiptHandle,
+			}); err != nil {
+				c.errorHandler.Handle(ctx, err)
+				c.errorEncoder(ctx, err, msg, c.sqsClient)
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// handleSingleMessage handles a single SQS message.
+func (c Consumer) handleSingleMessage(ctx context.Context, msg *sqs.Message, leftMsgs *[]*sqs.Message) error {
+	req, err := c.dec(ctx, msg)
+	if err != nil {
+		c.errorHandler.Handle(ctx, err)
+		c.errorEncoder(ctx, err, msg, c.sqsClient)
+		return err
+	}
+
+	response, err := c.e(ctx, req)
+	if err != nil {
+		c.errorHandler.Handle(ctx, err)
+		c.errorEncoder(ctx, err, msg, c.sqsClient)
+		return err
+	}
+
+	responseMsg := sqs.SendMessageInput{}
+	for _, f := range c.after {
+		ctx = f(ctx, msg, &responseMsg, leftMsgs, c.leftMsgsMux)
+	}
+
+	if !c.wantRep(ctx, msg) {
+		return nil
+	}
+
+	if err := c.enc(ctx, &responseMsg, response); err != nil {
+		c.errorHandler.Handle(ctx, err)
+		c.errorEncoder(ctx, err, msg, c.sqsClient)
+		return err
+	}
+
+	if _, err := c.sqsClient.SendMessageWithContext(ctx, &responseMsg); err != nil {
```

Comment:
I will introduce ResponsePublisher which it will responsible to publish the response or just run no operation func. How about that?

Reply:
I don't see what that would improve over my implementation. If you think it really does please help me understand.
I have a WantReplyFunc that can be either based on a message attribute or simply return false or true all the time.
Depending on the output of this func, I proceed to the rest of the function which is in charge of encoding and sending a response or not.

I prefere my approach because I find it closer to gokit's approach (in NATS transport for example) where you have a dedicated encoding response function.